### PR TITLE
Tutorials by cat

### LINKED ISSUE #72: Query for tutorials by category

### PR DESCRIPTION
**Pull request by sendusyourbones**
closes #72 

Add query to get all tutorials that belong to a specific category.
Update tutorial resolvers to use function declarations instead of arrow functions.

To test, `npm run seed` and then `npm start`. Go to `localhost:3001/graphql` in the browser and test the `tutorialsByCategory` query (first query categories to get a category ID).

### CODE REVIEW

File: server/schemas/tutorialDefs.js [hunk 1]
```diff
@@ -39,22 +40,32 @@ const tutorialResolvers = {
...
+        throw new Error(`Failed to fetch tutorials by category: ${error.message}`);
+      }
+    },
   },
+
   Mutation: {
     // Add a tutorial
-    addTutorial: async (parent, { title, overview, thumbnail, categories, teacher }) => {
+    addTutorial: async function(parent, { title, overview, thumbnail, categories, teacher }) {
       return await Tutorial.create({ title, overview, thumbnail, categories, teacher });
```
Comment (rileydoneil) [inline server/schemas/tutorialDefs.js]: try/catch

File: server/schemas/tutorialDefs.js [hunk 2]
```diff
@@ -78,10 +89,10 @@ const tutorialResolvers = {
     },
 
     // Delete a tutorial
-    deleteTutorial: async (parent, { _id }) => {
+    deleteTutorial: async function(parent, { _id }) {
       return await Tutorial.findByIdAndDelete(_id);
```
Comment (rileydoneil) [inline server/schemas/tutorialDefs.js]: try/catch

File: server/schemas/tutorialDefs.js [hunk 3]
```diff
@@ -39,22 +40,32 @@ const tutorialResolvers = {
     },
 
     // Get a single tutorial by ID
-    tutorial: async (parent, { _id }) => {
+    tutorial: async function(parent, { _id }) {
       return await Tutorial.findById(_id)
```
Comment (rileydoneil) [inline server/schemas/tutorialDefs.js]: try/catch

File: server/schemas/tutorialDefs.js [hunk 4]
```diff
@@ -30,7 +31,7 @@ const tutorialTypeDefs = gql`
 const tutorialResolvers = {
   Query: {
     // Get all tutorials
-    tutorials: async () => {
+    tutorials: async function() {
       return await Tutorial.find({})
```
Comment (rileydoneil) [inline server/schemas/tutorialDefs.js]: try/catch

**Review by rileydoneil — COMMENTED** [review]
Code looks great and performs well! I noticed that if you grab a tutorial using tutorialsbyCat that the category names of the returned tutorials are null. This is technically a non-issue just thought someone else should know. Great job!!

File: server/schemas/tutorialDefs.js (hunk 1 above)
Comment (sendusyourbones) [reply server/schemas/tutorialDefs.js:64]: I'm adding try/catch in a future branch/PR, thanks!

File: server/schemas/tutorialDefs.js (hunk 2 above)
Comment (sendusyourbones) [reply server/schemas/tutorialDefs.js:93]: I'm adding try/catch in a future branch/PR, thanks!

File: server/schemas/tutorialDefs.js (hunk 3 above)
Comment (sendusyourbones) [reply server/schemas/tutorialDefs.js:44]: I'm adding try/catch in a future branch/PR, thanks!

File: server/schemas/tutorialDefs.js (hunk 4 above)
Comment (sendusyourbones) [reply server/schemas/tutorialDefs.js:35]: I'm adding try/catch in a future branch/PR, thanks!

**Comment by sendusyourbones** [comment]
@rileydoneil Ah yeah that can be resolved by adding a `populate` statement to the query so that the category data is fully populated (like what I did in the `tutorials` query). It's not super clear to me in which cases we want to do that and which ones we don't. Want me to add it, or we can wait to see what happens on the front-end and add it later if needed?

**Comment by UserOlena** [comment]
It does find the tutorial by category, but for some reason query only includes IDs of fields with array but doesn't include all other field 
Screenshot 2023-07-18 at 2 54 41 PM

**Comment by UserOlena** [comment]
Additionally, when I include totalDuration field it crashes

Screenshot 2023-07-18 at 2 54 58 PM

**Comment by sendusyourbones** [comment]
> It does find the tutorial by category, but for some reason query only includes IDs of fields with array but doesn't include all other field Screenshot 2023-07-18 at 2 54 41 PM

See my comment to Riley above

File: server/schemas/tutorialDefs.js (hunk 2 above)
Comment (sendusyourbones) [reply server/schemas/tutorialDefs.js]: Added

File: server/schemas/tutorialDefs.js (hunk 4 above)
Comment (sendusyourbones) [reply server/schemas/tutorialDefs.js]: Added

File: server/schemas/tutorialDefs.js (hunk 1 above)
Comment (sendusyourbones) [reply server/schemas/tutorialDefs.js]: Added

File: server/schemas/tutorialDefs.js (hunk 3 above)
Comment (sendusyourbones) [reply server/schemas/tutorialDefs.js]: Added

**Comment by sendusyourbones** [comment]
@UserOlena @rileydoneil Missing data, `try/catch`, and `totalDuration` issue resolved with latest commit.

**Review by UserOlena — COMMENTED** [review]
It returns data for all the fields now

Screenshot 2023-07-18 at 3 52 34 PM

but I have a question for clarification just in case I do not understand it correctly.
When I run query for the `tutorialsByCategory` and include all the fields it can possibly return, because in the `tutorialDefs.js` `tutorialsByCategory` returns
`tutorialsByCategory(categoryId: ID!): [Tutorial]`
and `Tutorial` model has this field
```
type Tutorial {
    _id: ID!
    title: String
    overview: String
    thumbnail: String
    categories: [Category]
    teacher: [User]
    lessons: [Lesson]
    reviews: [Review]
    totalDuration: Int
  }
```

and it looks to me that it creates an infinity loop, because 
`tutorialsByCategory`  returns `[Tutorial]` and `Tutorial` returns `Teacher` and `Teacher` returns a `Tutorial` and so on. Should the `Teacher` field be excluded in this case ot a `Tutorial` field inside the `Teacher`?
Because, technically I couldn't even run the whole query until I removed a `Tutorial` field from the `Teacher`

**Comment by sendusyourbones** [comment]
I do not think we should remove `teacher` from `tutorial`. In the case that we are fetching a tutorial and getting its teacher, we don't care about the tutorials the teacher is enrolled in so we would not ask for the `tutorials` field on the teacher. On the front-end when we run the query, we specify which fields we want to get back, so we would make sure not to include `tutorials` on `teacher`.

**Comment by UserOlena** [comment]
> I do not think we should remove `teacher` from `tutorial`. In the case that we are fetching a tutorial and getting its teacher, we don't care about the tutorials the teacher is enrolled in so we would not ask for the `tutorials` field on the teacher. On the front-end when we run the query, we specify which fields we want to get back, so we would make sure not to include `tutorials` on `teacher`.

Yes, I understand that we shouldn't remove Teacher from the Tutorial. i meant to remove Tutorial field from Teacher on the `tutorialsByCategory` query

**Review by UserOlena — APPROVED** [review]

**Comment by UserOlena** [comment]
Works great!

Screenshot 2023-07-18 at 6 32 18 PM

**Pull request merged**